exporters/datadog/lib/opentelemetry/exporters/datadog/datadog_span_processor.rb

```diff
@@ -0,0 +1,199 @@
+# frozen_string_literal: true
+
+# Copyright 2019 OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+require 'timeout'
+
+module OpenTelemetry
+  module Exporters
+    module Datadog
+      # Implementation of the duck type SpanProcessor that batches spans
+      # exported by the SDK into complete traces then pushes them
+      # to the exporter pipeline.
+      #
+      # All spans reported by the SDK implementation are first added to a
+      # synchronized in memory trace storage (with a {max_queue_size}
+      # maximum size, of trace size {max_trace_size} after the size of
+      # either is reached spans are dropped). When traces are designated
+      # as "complete" they're added to a queue that is exported every
+      # schedule_delay_millis to the exporter pipeline in batches of
+      # completed traces. The datadog writer and transport supplied
+      # to the exporter handle the bulk of the timeout and retry logic.
+      class DatadogSpanProcessor
+        SCHEDULE_DELAY_MILLIS = 3_000
+        MAX_QUEUE_SIZE = 2048
+        MAX_TRACE_SIZE = 1024
+        private_constant(:SCHEDULE_DELAY_MILLIS, :MAX_QUEUE_SIZE, :MAX_TRACE_SIZE)
+
+        def initialize(exporter:,
+                       schedule_delay_millis: SCHEDULE_DELAY_MILLIS,
+                       max_queue_size: MAX_QUEUE_SIZE,
+                       max_trace_size: MAX_TRACE_SIZE)
+          raise ArgumentError if max_trace_size > max_queue_size
+
+          @exporter = exporter
+          @mutex = Mutex.new
+          @condition = ConditionVariable.new
+          @keep_running = true
+          @delay_seconds = schedule_delay_millis / 1000.0
+          @max_queue_size = max_queue_size
+          @max_trace_size = max_trace_size
+          @spans = []
+          @thread = Thread.new { work }
+
+          @traces = {}
+          @traces_spans_count = {}
+          @traces_spans_ended_count = {}
+          @check_traces_queue = []
+          @_spans_dropped = false
+        end
+
+        # datadog trace-agent endpoint requires a complete trace to be sent
+        # threadsafe may block on lock
+        def on_start(span)
+          context = span.context
+          trace_id = context.trace_id
+
+          lock do
+            if all_spans_count(traces_spans_count) >= max_queue_size
+              OpenTelemetry.logger.warn('Max spans for all traces, spans will be dropped')
```

Comment:
One alternative way to handle a full queue is to drop a random completed trace (dropping a full trace, not a single span). We currently do that in the [`ddtrace` gem](https://github.com/DataDog/dd-trace-rb/blob/c6fbf2410a60495f1b2d8912bf7ea7dc63422141/lib/ddtrace/buffer.rb#L34-L36).

It allows for a more fair usage of the queue when under stress load, and will create proportional representation of code paths being instrumented at stress time.

Reply:
that's a good idea, i've updated the processor to drop a random unfinished trace in this case